FW/FreeRTOS/Queue.hpp

```diff
@@ -0,0 +1,88 @@
+#ifndef _FREERTOS_QUEUE_HPP
+#define _FREERTOS_QUEUE_HPP
+
+#include "FreeRTOS.h"
+#include "queue.h"
+
+
+namespace FreeRTOS
+{
+    template<class T, size_t size>
+    class Queue {
+    public:
+        Queue() {
+            //m_handle = xQueueCreate(size, sizeof(T));
+            m_handle = xQueueCreateStatic(size, sizeof(T), m_buffer, &m_staticQueue);
+        }
+
+        ~Queue() {
+            vQueueDelete(m_handle);
+        }
+
+
+        /**
+         * @brief Add item to queue, if queue is full, thread will be paused until there
+         *        is room or timeout elapses. Do not use this within an interrupt, use
+         *        EnqueueFromInterrupt() instead. Will return true if item successfully
+         *        was added to queue, or false if timed out
+         * 
+         * @param item Item to add to queue
+         * @param timeout Timeout in milliseconds (max MaxTimeout)
+         * @return true Item was added to queue
+         * @return false Queue is full and timed out
+         */
+        bool Enqueue(T *item, const uint32_t timeout = MaxTimeout) {
+            TickType_t ticks = pdMS_TO_TICKS(timeout > MaxTimeout ? MaxTimeout : timeout);
+            return xQueueSend(m_handle, item, ticks) == pdTRUE;
+        }
+
+
+        /**
+         * @brief Add item to queue from an interrupt
+         * 
+         * @param item Item to add to queue
+         * @return true Item was added to queue
+         * @return false Queue is full
+         */
+        bool EnqueueFromInterrupt(T *item) {
+            return xQueueSendFromISR(m_handle, item, NULL) == pdTRUE;
+        }
+
+
+        /**
+         * @brief Get item from the queue. If queue is empty thread will be paused until
+         *        there is an item to get or until timeout has elapsed
+         * 
+         * @param item Pointer to the item to get
+         * @param timeout Timeout in milliseconds (max MaxTimeout)
+         * @return true Item was retrieved
+         * @return false Queue was empty and timed out
+         */
+        bool Dequeue(T *item, const uint32_t timeout = MaxTimeout) {
+            TickType_t ticks = pdMS_TO_TICKS(timeout > MaxTimeout ? MaxTimeout : timeout);
+            return xQueueReceive(m_handle, item, ticks) == pdTRUE;
+        }
+
+
+        /**
+         * @brief Number of milliseconds for no timeout (defined as 24 H)
+         * 
+         */
+        static constexpr uint32_t MaxTimeout = pdMS_TO_TICKS(86400000);
```

Comment:
I'd vote for following the convention of consts in all-caps, i.e. MAX_TIMEOUT etc.

Reply:
Playing around with this... I actually think it's more readable this way. 

Heard a talk recently about writing code that you more or less can read as a English text. Avoiding abbreviations and "shouty" all-caps defines. One of the reasons being that you involve more parts of your brain while reading the code, not only the programming part but also the lingual. And that allows you to more quickly understand the code... Not sure what I think yet, but it's funny playing around with in a  project like this.

Reply:
OK, hadn't really thought of it like that. I'm very used to the constant shouting (pun intended). Perhaps it's a foregone conclusion that it's good to separate variables from constants in the naming scheme, not sure though. Also, I guess they still are if you snake-case variable names...